Merge `wcs-migration` into `trunk`

Ferdev opened the pull request:
## Description
This PR merges all work done in the `wcs-migration` branch to trunk, to prepare for the future release of the Migration Flow project.

### Related issue(s)
N/A

### Steps to reproduce & screenshots/GIFs
N/A

### Checklist

<!-- All testable code should have tests. -->
- [ ] unit tests

<!-- An entry in the changelog file is always required -->
- [ ] `changelog.txt` entry added
- [ ] `readme.txt` entry added

samnajian committed:
Define Feature Announcement component and open it on click [122] (#2743)

* Define Feature Announcement component and add the content

* Open the feature announcement modal when user clicks on Create shipping label

* Make the copy translatable

* Aknowledge eligibility when opening the feature announcement modal

* Add buttons and polish styling

* Disable the modal for the time being

* Add missing text and update styling

* Remove redundant argument in selector

* Add modal left column not taking 100% of the height

* Remove useCallback from stub function

* Revert code formatting back

samnajian committed:
Fix/adjust how feature announcement looks on smaller screens (#2746)

* Fix how Feature Announcement component looks on smaller screens

* Remove excess space in closing tag

Commit:
Install and activate a list of plugins during migrations (#2745)

* Add @wordpress/api-fetch package

* Created a place holder button to test plugin activation

* Update shrinkwrap since we installed api-fetch

* Use hello-dolly to test install and activate

* Use native fetch instead of apiFetch

* Remove apiFetch package

* Fix < /p> to </p>

* Bind activation and installation to the update button

* Deactivate WCS&T after installing Woo Shipping and Woo Tax

* Disable is_eligible_for_migration

Don't enable migration yet.

* Redirect to plugins after installation

* Fix deactivateWCSTPluginAPICall to return promise

* Use relative path for the actions in feature-announcement

* Add adminPluginPath in the js global config.

* Execute each task in steps and throw exception if any fails

Commit:
Fix fill-rule to fillRule and clip-rule to clipRule (#2748)

Commit:
Set a flag to indicate migration has started or has completed (#2747)

* Add connect/migration-flag endpoint

* Set a flag "wcshipping_migrated" to indicate it is migrated

* is_eligible_for_migration() now checks for the config as well

* Add migration done API call to the update button

* Update migration flag to migration state

* Update space to tabs

* Call API to update migration status to "started"

* Update migration state to completed in plugin deactivation hook

* Default is_eligible_for_migration to return false

* Update API constants to something like an enum instead

* PHP cs fixes

* Return 200 if option updated, 304 otherwise

* Remove redundancy in condition

* Inline code comment can not be a DocBlock

* Remove redundant class_exists() check for the enum class

* Return 500 if update_option failed

Commit:
Add admin notice on orders page for WCShipping migration (#2751)

* Use current_screen hook to trigger the add admin notice hook

* Add dimissable

* Added a button for "Upgrade now"

* Update "Confirm update" style

* Fix spaces

* Fix indentation

* Fix whitespace for linting

* Screen ID not found if using HPOS, fixed

* Update learn more link

* Add a new function is_on_order_list_page() to check order list page

Commit:
Merge branch 'trunk' into wcs-migration

Ferdev committed:
Get the migration banner info from the connect server (#2752)

* Add note to README with a fix for `npm i` getting stuck

* Retrieves the migration banner info from WooCommerce Connect Server

* Address PR comments

Ferdev committed:
Get the flag to display the migration banner from the connect server (#2755)

* Refactor code checking whether or not to display a migration banner

* Address PR comments

samnajian committed:
Change dismiss icon and behaviour [shipping-140] (#2758)

* Add is-dismissable and button label from server

* Register separate stylesheet and add stylings for the banner and buttons

* Add functionality to dismiss and remember dismissal

Commit:
Add state machine to drive migration states (#2757)

* Add state machine to drive migration states

* Rename state to better describe its name

* Refactor the API call out so the runNext function is decouple from API status check

* Update migration state enums to include all states in the machine

* Store the state in the wcshipping_migration_state option.

This allows us to remember where we were and we can always restart the
migration based on the last known state.

* Remove comment.

* Pass the wcshipping_migration_state option into the FeatureAnnouncement react component

* Fix a bug where restarting from an error state didn't proceed to the next state

* Return 304 if state is not modified instead of throwing 4xx.

* Fix 'wcshippingMigrationState' is already declared in the upper scope error

* Refactor the state machine logic into its own module

* Use wcs-client as path and re-enable redirection

* Added comments to the migration-runner module

* Apply suggestions from code review

Lint fixes.

Co-authored-by: Fernando Espinosa <Ferdev@users.noreply.github.com>

* Wait for the migration to finish before setting isUpdating back to false

* Return the fetch Response object instead

* Added a function to check if state exists before using it.

Throw exception otherwise.

* Update client/components/migration/migration-runner.js

* Update client/components/migration/migration-runner.js

* Update client/components/migration/migration-runner.js

---------

Co-authored-by: Fernando Espinosa <Ferdev@users.noreply.github.com>

Ferdev committed:
Add a "deactivated" message to the plugins list entry (#2759)

* Add a "deactivated" message to the plugins list entry

* Update woocommerce-services.php

Co-authored-by: Gerhard Potgieter <potgieterg@gmail.com>

* Address PR comments

* Remove redundant `esc_html`

* Customize the after_plugin_row action for this plugin only

* Fix background color of deactivated message

---------

Co-authored-by: Gerhard Potgieter <potgieterg@gmail.com>

Commit:
Load the FeatureAnnouncement modal to the order listing page (#2756)

* Load a new javascript file when migration admin notice shows up

* Load redux store for admin notice

The redux store inlcudes only the label shipping's reducer and the initial states.

* semi-colon gets rendered out. Removing.

* Fix style and setup initial data from the PHP's side

* Use wcs-client alias for webpack otherwise order list page can't find it

* Add baseURL and redirect URL through the proper enqueued scripts

* Move inline styles to the migration scss

* Clicking "Confirm update" will start the migration process

* Remove phpcs:ignore because it does have a version

* Use ID as the selector for the modal update button instead.

* Fix merge conflict issues with the existing css and js overlaps

* Change testing 10s back to 3 * 24 * 60 * 60

* Fix admin notice dimissible button and cleaned up jquery

* After clicking dimiss, remove itself from the DOM until 3 days later

* Add EOL to the js file.

* Change headers to a callback to retrieve getNonce() after dependencies are loaded

* Refactor cookie function to window.wpCookies

Ferdev committed:
Add tracking data for the migration process (#2760)

* Add tracking data for the migration process

* Remove debugging output

* Add update value to tracked data

Commit:
Clicking "Maybe later" will dismiss feature announcement for 3 days (#2762)

* Clicking "Maybe later" will dismiss the modal for 3 days

* Do not display modal if it's previously dismissed

* Don't show the modal if the page isn't refreshed

* EOL in new file

* Apply suggestions from code review

Lint fixes.

Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

---------

Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

Ferdev committed:
Save transients after a migration has been completed (#2763)

* Save transients after a migration has been completed

* Use general options instead of WooCommerce Shipping & Tax's options

* Cast the migration state to int now that is saved as a global option

* Remove wcshipping_migration_state as option name

Ferdev committed:
Add missing tracking data for the migration process (#2765)

* Update tracking data when the migration is completed

* Move plugin entry deactivated message to WooCommerce Shipping

* Add missing imports for tracks in a static method

Commit:
Remove db migration and deactivation (#2767)

* Skip db and deactivating plugin and finish the migration once activation is done

* Remove woocommerce-tax from the download and activation list

* Because WCS&T is no longer deactivated, we can update the migration state when it quits

* No longer need to rely on plugin_deactivation() to update the final migration state

Ferdev committed:
Adapt the migration banner popup to reusing WCS&T (#2769)

* Change copy in migration popup

* Improve copy in migration popup

samnajian committed:
Provide public path to js entry files (#2770)

samnajian committed:
Display migration banner on all WC settings page [shipping-152] (#2771)

* Show upgrade banner on shipping settings page as well

* Make sure clicking the confirm upgrade button doesn't refresh the page

* Show update banner on all settings pages

Ferdev committed:
Rebase `wcs-migration` branch onto `trunk` (#2783)

* Tweak/issue 2750 - WordPress 6.6 and WooCommerce 9.0 Compatibility (#2754)

* Bump tested up to version

* add changelog

* update readme file

* woorelease: Product version bump update

* Revert "woorelease: Product version bump update"

This reverts commit 122c7b3af955053f0fb0ea5ac01ec2264a4ed88f.

* woorelease: Product version bump update

* Revert "woorelease: Product version bump update"

This reverts commit 642db4fa3010a71d3bbfa4b52f7edfd99b36fcd7.

* woorelease: Product version bump update

* Fix/issue 2712 - Make The Tariff Number Field a Required Field when the Destination is Within EU (#2764)

* add condition for EU tariff number

* add changelog

* edit the tariff number error text

* edit changelog

* woorelease: Product version bump update

* Conditionally load shipping functionality based on if WC Shipping is active (#2761)

* Rough take

* Do not change the name, this is still called WooCommerce Shipping & Tax

* Set `wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax` hook to indicate we will handle the coexistence

* Remove WC_Connect_Note_DHL_Live_Rates_Available notice todo

My concern was about the class not existing and thereby creating PHP fatal errors, but we require the file right before using it, so there's no issue even if shipping has not been generally initiated.

* Remove outdated todos

* Status page should respect shipping conditional loading (#2776)

* Conditionally hide shipping related status data

* Satisfy tests

* Introduce new "wc_services_will_disable_shipping_logic" hook (#2775)

* Reimplement logic that completely blocks WCS&T logic from initiating

* Introduce new hook for conditional loading

---------

Co-authored-by: Gerhard <potgieterg@gmail.com>

* Add changelog for WC Shipping compatibility (#2779)

* Conditionally shows legacy reports shipping tab (#2780)

* woorelease: Product version bump update

* Define Feature Announcement component and open it on click [122] (#2743)

* Define Feature Announcement component and add the content

* Open the feature announcement modal when user clicks on Create shipping label

* Make the copy translatable

* Aknowledge eligibility when opening the feature announcement modal

* Add buttons and polish styling

* Disable the modal for the time being

* Add missing text and update styling

* Remove redundant argument in selector

* Add modal left column not taking 100% of the height

* Remove useCallback from stub function

* Revert code formatting back

* Fix/adjust how feature announcement looks on smaller screens (#2746)

* Fix how Feature Announcement component looks on smaller screens

* Remove excess space in closing tag

* Install and activate a list of plugins during migrations (#2745)

* Add @wordpress/api-fetch package

* Created a place holder button to test plugin activation

* Update shrinkwrap since we installed api-fetch

* Use hello-dolly to test install and activate

* Use native fetch instead of apiFetch

* Remove apiFetch package

* Fix < /p> to </p>

* Bind activation and installation to the update button

* Deactivate WCS&T after installing Woo Shipping and Woo Tax

* Disable is_eligible_for_migration

Don't enable migration yet.

* Redirect to plugins after installation

* Fix deactivateWCSTPluginAPICall to return promise

* Use relative path for the actions in feature-announcement

* Add adminPluginPath in the js global config.

* Execute each task in steps and throw exception if any fails

* Fix fill-rule to fillRule and clip-rule to clipRule (#2748)

* Set a flag to indicate migration has started or has completed (#2747)

* Add connect/migration-flag endpoint

* Set a flag "wcshipping_migrated" to indicate it is migrated

* is_eligible_for_migration() now checks for the config as well

* Add migration done API call to the update button

* Update migration flag to migration state

* Update space to tabs

* Call API to update migration status to "started"

* Update migration state to completed in plugin deactivation hook

* Default is_eligible_for_migration to return false

* Update API constants to something like an enum instead

* PHP cs fixes

* Return 200 if option updated, 304 otherwise

* Remove redundancy in condition

* Inline code comment can not be a DocBlock

* Remove redundant class_exists() check for the enum class

* Return 500 if update_option failed

* Add admin notice on orders page for WCShipping migration (#2751)

* Use current_screen hook to trigger the add admin notice hook

* Add dimissable

* Added a button for "Upgrade now"

* Update "Confirm update" style

* Fix spaces

* Fix indentation

* Fix whitespace for linting

* Screen ID not found if using HPOS, fixed

* Update learn more link

* Add a new function is_on_order_list_page() to check order list page

* Get the migration banner info from the connect server (#2752)

* Add note to README with a fix for `npm i` getting stuck

* Retrieves the migration banner info from WooCommerce Connect Server

* Address PR comments

* Get the flag to display the migration banner from the connect server (#2755)

* Refactor code checking whether or not to display a migration banner

* Address PR comments

* Change dismiss icon and behaviour [shipping-140] (#2758)

* Add is-dismissable and button label from server

* Register separate stylesheet and add stylings for the banner and buttons

* Add functionality to dismiss and remember dismissal

* Add state machine to drive migration states (#2757)

* Add state machine to drive migration states

* Rename state to better describe its name

* Refactor the API call out so the runNext function is decouple from API status check

* Update migration state enums to include all states in the machine

* Store the state in the wcshipping_migration_state option.

This allows us to remember where we were and we can always restart the
migration based on the last known state.

* Remove comment.

* Pass the wcshipping_migration_state option into the FeatureAnnouncement react component

* Fix a bug where restarting from an error state didn't proceed to the next state

* Return 304 if state is not modified instead of throwing 4xx.

* Fix 'wcshippingMigrationState' is already declared in the upper scope error

* Refactor the state machine logic into its own module

* Use wcs-client as path and re-enable redirection

* Added comments to the migration-runner module

* Apply suggestions from code review

Lint fixes.

Co-authored-by: Fernando Espinosa <Ferdev@users.noreply.github.com>

* Wait for the migration to finish before setting isUpdating back to false

* Return the fetch Response object instead

* Added a function to check if state exists before using it.

Throw exception otherwise.

* Update client/components/migration/migration-runner.js

* Update client/components/migration/migration-runner.js

* Update client/components/migration/migration-runner.js

---------

Co-authored-by: Fernando Espinosa <Ferdev@users.noreply.github.com>

* Add a "deactivated" message to the plugins list entry (#2759)

* Add a "deactivated" message to the plugins list entry

* Update woocommerce-services.php

Co-authored-by: Gerhard Potgieter <potgieterg@gmail.com>

* Address PR comments

* Remove redundant `esc_html`

* Customize the after_plugin_row action for this plugin only

* Fix background color of deactivated message

---------

Co-authored-by: Gerhard Potgieter <potgieterg@gmail.com>

* Load the FeatureAnnouncement modal to the order listing page (#2756)

* Load a new javascript file when migration admin notice shows up

* Load redux store for admin notice

The redux store inlcudes only the label shipping's reducer and the initial states.

* semi-colon gets rendered out. Removing.

* Fix style and setup initial data from the PHP's side

* Use wcs-client alias for webpack otherwise order list page can't find it

* Add baseURL and redirect URL through the proper enqueued scripts

* Move inline styles to the migration scss

* Clicking "Confirm update" will start the migration process

* Remove phpcs:ignore because it does have a version

* Use ID as the selector for the modal update button instead.

* Fix merge conflict issues with the existing css and js overlaps

* Change testing 10s back to 3 * 24 * 60 * 60

* Fix admin notice dimissible button and cleaned up jquery

* After clicking dimiss, remove itself from the DOM until 3 days later

* Add EOL to the js file.

* Change headers to a callback to retrieve getNonce() after dependencies are loaded

* Refactor cookie function to window.wpCookies

* Add tracking data for the migration process (#2760)

* Add tracking data for the migration process

* Remove debugging output

* Add update value to tracked data

* Clicking "Maybe later" will dismiss feature announcement for 3 days (#2762)

* Clicking "Maybe later" will dismiss the modal for 3 days

* Do not display modal if it's previously dismissed

* Don't show the modal if the page isn't refreshed

* EOL in new file

* Apply suggestions from code review

Lint fixes.

Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

---------

Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

* Save transients after a migration has been completed (#2763)

* Save transients after a migration has been completed

* Use general options instead of WooCommerce Shipping & Tax's options

* Cast the migration state to int now that is saved as a global option

* Remove wcshipping_migration_state as option name

* Add missing tracking data for the migration process (#2765)

* Update tracking data when the migration is completed

* Move plugin entry deactivated message to WooCommerce Shipping

* Add missing imports for tracks in a static method

* Remove db migration and deactivation (#2767)

* Skip db and deactivating plugin and finish the migration once activation is done

* Remove woocommerce-tax from the download and activation list

* Because WCS&T is no longer deactivated, we can update the migration state when it quits

* No longer need to rely on plugin_deactivation() to update the final migration state

* Adapt the migration banner popup to reusing WCS&T (#2769)

* Change copy in migration popup

* Improve copy in migration popup

* Provide public path to js entry files (#2770)

* Display migration banner on all WC settings page [shipping-152] (#2771)

* Show upgrade banner on shipping settings page as well

* Make sure clicking the confirm upgrade button doesn't refresh the page

* Show update banner on all settings pages

---------

Co-authored-by: Luthfi Bintoro <iyut85@yahoo.com>
Co-authored-by: Sam Najian <sam.najian@automattic.com>
Co-authored-by: Harris Wong <harris.wong@automattic.com>
Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>
Co-authored-by: Gerhard <potgieterg@gmail.com>
Co-authored-by: Sam Najian <dev.samnajian@gmail.com>

Ferdev committed:
Fix error launching the migration popup (#2784)

samnajian committed:
Upgrade jetpack packages to version compatible with WCShipping (#2785)

samnajian committed:
Use latest WP version for e2e tests

Ferdev committed:
Merge `trunk` into `wcs-migration` (#2795)

* Tweak/issue 2750 - WordPress 6.6 and WooCommerce 9.0 Compatibility (#2754)

* Bump tested up to version

* add changelog

* update readme file

* woorelease: Product version bump update

* Revert "woorelease: Product version bump update"

This reverts commit 122c7b3af955053f0fb0ea5ac01ec2264a4ed88f.

* woorelease: Product version bump update

* Revert "woorelease: Product version bump update"

This reverts commit 642db4fa3010a71d3bbfa4b52f7edfd99b36fcd7.

* woorelease: Product version bump update

* Fix/issue 2712 - Make The Tariff Number Field a Required Field when the Destination is Within EU (#2764)

* add condition for EU tariff number

* add changelog

* edit the tariff number error text

* edit changelog

* woorelease: Product version bump update

* Conditionally load shipping functionality based on if WC Shipping is active (#2761)

* Rough take

* Do not change the name, this is still called WooCommerce Shipping & Tax

* Set `wc_services_will_handle_coexistence_with_woo_shipping_and_woo_tax` hook to indicate we will handle the coexistence

* Remove WC_Connect_Note_DHL_Live_Rates_Available notice todo

My concern was about the class not existing and thereby creating PHP fatal errors, but we require the file right before using it, so there's no issue even if shipping has not been generally initiated.

* Remove outdated todos

* Status page should respect shipping conditional loading (#2776)

* Conditionally hide shipping related status data

* Satisfy tests

* Introduce new "wc_services_will_disable_shipping_logic" hook (#2775)

* Reimplement logic that completely blocks WCS&T logic from initiating

* Introduce new hook for conditional loading

---------

Co-authored-by: Gerhard <potgieterg@gmail.com>

* Add changelog for WC Shipping compatibility (#2779)

* Conditionally shows legacy reports shipping tab (#2780)

* woorelease: Product version bump update

* Update readme.txt to indicate that a new Woo Shipping plugin is available for download. (#2789)

* Update changelog

* Remove changelog and updated copy

* Make sure docker-compose command exists (#2793)

---------

Co-authored-by: Luthfi Bintoro <iyut85@yahoo.com>
Co-authored-by: Sam Najian <sam.najian@automattic.com>
Co-authored-by: Harris Wong <harris.wong@automattic.com>
Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>
Co-authored-by: Gerhard <potgieterg@gmail.com>
Co-authored-by: Sam Najian <dev.samnajian@gmail.com>

Ferdev committed:
Add a new widget in settings to start the migration (#2796)

* Add a new widget in settings to start the migration

* Add tracks event for migration settings widget

Ferdev committed:
Merge branch 'trunk' into wcs-migration

harriswong reviewed (approved):
I pulled this feature branch and did a quick smoke test with WCShipping disabled. I can confirm the migration modal. I clicked through the modal and I was able to get 2 plugins activated
![image](https://github.com/user-attachments/assets/dc79a857-38f7-47d2-8f9f-35fbfe6e7403)

I can also get the new WCShipping banner to show inside an order.

**Question**
Double checking if there are any tasks we need to do before merging this? Otherwise, LGTM 👍

Ferdev commented:
> Double checking if there are any tasks we need to do before merging this? Otherwise, LGTM 👍

Thanks for the question Harris! We can totally merge this now, but it won't be enabled for regular accounts until the migration flag in WooCommerce Connect Server is activated for `production` env. Other than that, I see no issues merging it now.

Ferdev committed:
Update changelog and readme

kallehauge commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -0,0 +1,146 @@
+					<h2>
+						{translate('A new dedicated WooCommerce Shipping extension is now available')}
+					</h2>
+					<p>{translate('We\'ll ensure a smooth transition by providing the functionality to carry over all your settings and shipping labels when you update.')}</p>
```
Very late/last minute suggestion: The "providing the functionality to carry" part made me go back and re-read it, so perhaps a simple change like this would be OK last minute? 😄 

```suggestion
					<p>{translate('We will ensure a seamless transition by transferring of all your settings and shipping labels when you update.')}</p>
```

harriswong replied:
Looks good to me, i am committing your suggestion.

harriswong replied:
Actually, I will change it to "We will ensure a seamless transition by transferring ~~of~~ all your settings and shipping labels when you update."

harriswong replied:
Fixed in 71637e12d01a01cfeb932deebcfba170472d3e28

kallehauge commented on client/components/migration/feature-announcement.jsx:
```diff
@@ -0,0 +1,146 @@
+import { installAndActivatePlugins } from './migration-runner';
+import { TIME_TO_REMMEMBER_DISMISSAL_SECONDS } from './constants';
+
+const FeatureAnnouncement = ( { translate, isEligable, previousMigrationState, onClose } ) => {
```
I noticed that some of the padding around our "content" is showing a purple border. Should this be white?

![Screenshot 2024-08-20 at 16 55 02](https://github.com/user-attachments/assets/f9d7b2bb-da5c-478f-8ec7-54a259ca49f1)

harriswong replied:
Yes there shouldn't be a gap 0EcJhbcadV2uAqVrA8hb6a-fi-467_5757. I created a follow up for this https://github.com/Automattic/woocommerce-services/issues/2801.

kallehauge replied:
Thank you Harris! IMO it's not blocking and honestly could look like it's intentional, so I'm marking this comment as resolved ❤️ 

(CC: @kloon, courtesy ping since you're the DRI)

kallehauge commented on changelog.txt:
```diff
@@ -1,5 +1,8 @@
 *** WooCommerce Shipping & Tax Changelog ***
 
+= 2.7.1 - 2024-xx-xx =
+* Add - User interface components to perform the migration to the new WooCommerce Shipping plugin.
```
I read this as the functionality already existed since it's a UI comment - can be ignored 😄 

```suggestion
* Add - A new migration experience from this plugin to the newly released WooCommerce Shipping plugin.
```

harriswong replied:
Looks good to me, I am committing your suggestion.

kallehauge reviewed (approved):
Works as expected 👏  

I gave this a spin to test the whole flow while reviewing https://github.com/woocommerce/woocommerce-shipping/pull/667 and found a few minor things. None of them are blocking in any way!

_(Edit: I've not reviewed the code itself - purely communicating I had a successful experience from a manual testing point of view 😄)_

Commit:
Update changelog.txt copy

Update the copy to "A new migration experience from this plugin to the newly released WooCommerce Shipping plugin."

Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

Commit:
Update feature announcement copy

harriswong commented:
👋 Hi @kallehauge! I addressed the comments, please take a look again!

kallehauge reviewed (approved):
I'm a happy man, @harriswong! 😄 

Thank you for looking into my comments 🙇

kloon committed:
Do not test on SKU, https://github.com/woocommerce/woocommerce/issues/47476 changed sku behaviour in tests

kloon committed:
Update version for migration work
